### PR TITLE
#167313417 Users should be able to bookmark articles for later reading

### PR DESCRIPTION
**Pull request by Anguandia**
#### What does this PR do?

  Add bookmarks feature

#### Description of Task to be completed?

  - Implement basic create read update and delete bookmarks functionalities
  - Grouping of bookmarks in collections
  - Duplication of bookmarks

#### How should this be manually tested?

  With a base url of https://codepirates-ah-backend.herokuapp.com/api/v1/users for remote and http://localhost:port/api/v1/users for local testing, send the following requests the *requests are authenticated*:

  - **Create a bookmark** POST/bookmarks with a body a son object having properties of name and articleId. The articleId must be an integer, the name could be empty or any string
    *Possible outcomes:*
    - With a valid id of an existing article, a bookmark should be created with the supplied name if not empty or the article title appended with a timestamp if current article not bookmarked by current user before
    - Try creating the same bookmark again, changing alternately the name and articleId values, randomly or targeted and examine the possible outcomes

  - **Update bookmark** PATCH/bookmarks/*target-bookmark-name* supplying a json object having the desired new name as property name and the articleId. Try all possible combinations of these three parameters, random, repeating and specific to explore the different behaviours of the application

  - **Get a specific bookmark** GET/bookmarks/*target-bookmark-name* 

  - **Get all bookmarks of the current user** GET/bookmarks

  - **Delete a specific bookmark** DELETE/bookmarks/*target-bookmark-name*

  - **Delete all bookmarks of the current user** DELETE/bookmarks

  - **Add a bookmark to a collection** POST/bookmarks/collections with a body having the properties collection for collection name, name for bookmark name and articleId

  - **Change a collection's name** PATCH/bookmarks/collections/*target-collection-name* supplying the desired new name as value to a collection property in the request body

  - **View a specific collection** GET/bookmarks/collections/*target-collection-name*

  - **Get all of user's collections** GET/bookmarks/collections

  - **Delete a bookmark from a collection** DELETE/bookmarks/collections/*target-collection*/*target-bookmark*

  - **Delete a whole collection** DELETE/bookmarks/collections/*collection-name*

#### What are the relevant pivotal tracker stories?

  [finishes #167313417](https://www.pivotaltracker.com/story/show/167313417)
#### Screenshots (if appropriate)

Create bookmark
Screen Shot 2019-08-27 at 22 50 38

Change a bookmark's name
Screen Shot 2019-08-27 at 22 52 33

Get a specific bookmark
Screen Shot 2019-08-27 at 22 54 22

Delete all bookmarks
Screen Shot 2019-08-27 at 22 56 08

### CODE REVIEW

**Review by nkalyesubula — DISMISSED** [review]
LGTM

**Review by Mnickii — DISMISSED** [review]
LGTM

**Review by nshutijonathan — DISMISSED** [review]

File: src/controllers/bookmarks.controller.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    const { collection } = req.params;
+    await deleteCollection(collection);
+    return res.status(200).json({ message: `collection '${collection}' deleted` });
+  }
+
+  // delete from collection
+  static async unCollect(req, res) {
+    const { collection, name } = req.params;
+    await unCollect(collection, name, req.auth.id);
+    return res.status(200).json({
```
Comment (Mnickii) [inline src/controllers/bookmarks.controller.js]: should be 404 @Anguandia

File: src/controllers/bookmarks.controller.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      message: `bookmark '${bookmark.name}' found`,
+      data: bookmark
+    });
+  }
+
+  static async deleteUserBookMark(req, res) {
+    const bookmark = await getBookMarkName(req.auth.id, req.params.name);
+    bookmark.destroy();
+    return res.status(200).json({
```
Comment (Mnickii) [inline src/controllers/bookmarks.controller.js]: same here

File: src/controllers/bookmarks.controller.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      message: `bookmark '${req.params.name}' deleted`
+    });
+  }
+
+  static async deleteUserBookMarks(req, res) {
+    const reader = await checkItem(req.auth.id, 'user', ['articles']);
+    const deleted = await reader.removeArticles(reader.articles);
+    res.status(200).json({ message: `${deleted} bookmarks deleted` });
```
Comment (Mnickii) [inline src/controllers/bookmarks.controller.js]: and here

File: src/middlewares/bookmarks.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
+    if (bookmark) {
+      return res.status(200).json({
+        message: `bookmark '${name}' exists`, data: bookmark
+      });
+    }
+    if (existing && existing.name) {
+      data = existing;
+      data.newName = name;
+      return res.status(200).json({
```
Comment (Mnickii) [inline src/middlewares/bookmarks.js]: should be 400 or 409

**Review by Mnickii — COMMENTED** [review]
@Anguandia please refactor these codes too

**Review by minega25 — DISMISSED** [review]

File: src/middlewares/bookmarks.js (hunk 4 above)
Comment (Anguandia) [reply src/middlewares/bookmarks.js]: Done

**Review by minega25 — DISMISSED** [review]

File: src/models/bookmark.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+module.exports = (sequelize, DataTypes) => {
+    const BookMark = sequelize.define('BookMark', {
+        articleId: DataTypes.INTEGER,
+        userId: DataTypes.INTEGER,
+        name: DataTypes.STRING,
+        collection: DataTypes.STRING
+    }, {});
+    BookMark.associate = function (models) {
+        BookMark.belongsTo(models.Article, { foreignKey: 'articleId' })
```
Comment (habinezadalvan) [inline src/models/bookmark.js:10]: @Anguandia what do you think when BookMark and Article relationship has onUpdate: 'CASCADE', and
	      onDelete: 'CASCADE?

**Review by Mnickii — DISMISSED** [review]
LGMT

**Pull request merged**